proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+##### Tuples
+
+In the Swift language, tuples are "exploded" out in the calling convention. Each element of a tuple is passed as a separate parameter. C# has its own language-level tuple feature in the ValueTuple family of types. The Swift/.NET interop story could choose to automatically handle tuple types in a similar way for a `CallConvSwift` function call. However, processing the tuple types in the JIT/AOT compilers is complicated and expensive. It would be much cheaper to defer this work upstack to the language projection, which can split out the tuple into individual parameters to pass to the underlying function. The runtime and libraries teams could add analyzers to detect value-tuples being passed to `CallConvSwift` functions at compile time to help avoid pits of failure.
+
+##### Automatic Reference Counting and Lifetime Management
+
+Swift has a very strongly-defined lifetime and ownership model. This model is specified in the Swift ABI and is similar to Objective-C's ARC (Automatic Reference Counting) system. The Binding Tools for Swift tooling handles these explicit lifetime semantics with some generated Swift code. In the new Swift/.NET interop, management of these lifetime semantics will be done by the Swift projection and not by the raw calling-convention support.
+
+##### Structs
+
+Like .NET, Swift has both struct types and class types. Unlike .NET, Swift's struct types have strong lifetime semantics more similar to C++ types than .NET structs. At the Swift ABI layer, there are broadly three types of structs: "POD/Trivial" structs, "Bitwise Takable/Movable" structs, and non-bitwise movable structs. The [Swift documentation](https://github.com/apple/swift/blob/main/docs/ABIStabilityManifesto.md#layout-and-properties-of-types) covers these different kinds of structs. Let's look at how we could map each of these categories of structs into .NET.
+
+"POD/Trivial" structs have no memory management required and no special logic for copying/moving/deleting the struct instance. Structs of this category can be represented as C# structs with the same field layout.
+
+"Bitwise Takable/Movable" structs have some memory management logic and require calls to Swift's ref-counting machinery to maintain expected lifetimes. Structs of this category can be projected into C# as a struct. When creating this C# struct, we would semantically treat each field as a separate local, create the C# projection of it, and save this "local" value into a field in the C# struct.
```

Comment:
Should this rather be in "Swift to .NET Language Feature Projections" section? I do not think we would want to do this as part of the runtime support.

Reply:
As an FYI, in the early days of Binding Tools for Swift, I tried to do exactly this, making a distinction between blitable and non-blittable structs. It doubles the number of special cases and was a factory for bugs all along the line. In addition, this should not be limited to structs, but should be extended to enums, which are also value types. In addition to blitability issues, this brings up the packing of cases in swift enums which is not only non-trivial, it's not documented, subject to change and fragile (adding a case can change the packing). Further, value types in swift are semantically different from value types in C# in that code may get executed when they go out of scope. For all of these reasons, I chose to implement them as `IDisposable` classes with an opaque payload. In terms of passing them more efficiently, we have more options at our disposal. We could take a pinvoke like this:
```
internal extern void someCallWithAStruct (SwiftStruct st);
````
and reimplement it as this:
```
internal extern void someCallWithAStruct_exploded4w (ulong w0, ulong w1, ulong w2, ulong w4)
```
As a final however, this may be moot if we have access to the self pointer. This is because if we require `enable-library-evolution` then the swift compiler generates thunks that use pass by reference for the self pointer (if any) and IIRC, passes all structs by reference (making a copy on the stack).